### DISCUSSION TITLE
Solidus 3.0 Extensions Hackathon

### DISCUSSION THREAD

**Discussion by kennyadsl** [solidusio/solidus#3911] [Announcements]
Hello Solidus Community! 👋 

We need your help to make all the extensions compatible with the next major release of Solidus: **Solidus 3.0**.

🙏 🙏 🙏 🙏 🙏 🙏 🙏 🙏 🙏 🙏 

We are organizing a Hackathon to work together and get most of the extensions ready.

### When?

📆 Friday, February, 5th the whole day, no matter your timezone.

### Why we need this?

All 2.x minor releases gradually deprecated Solidus Core code by printing a deprecation message when that code is used. We know that all Solidus users that want to upgrade will have to remove the deprecated code from their applications but that's not enough, since any extension used would also need to be updated to support **also** the newly released Solidus version.

### How can you help?

You can help by taking care of one or more extensions and upgrading them. We will host a Hackaton to organize the work asynchronously, provide help and get the most out of the community effort. 

We will track the work using [this GitHub Project](https://github.com/orgs/solidusio-contrib/projects/1).

### Process

Here's the steps needed to participate:

<details>
<summary><b>1. Ask to join the GitHub Project Board</b></summary>

Let us know that you want to join the Hackathon using the comments below and we will add you to the [3.0 Preppers Board](https://github.com/orgs/solidusio-contrib/projects/1/). Now you can move cards around in the GitHub Project.

</details>

<details>
<summary><b>2. Choose one extension from the TODO list</b></summary>

Try to choose an extension that you are familiar with or one that you use in your project.
</details>

<details>
<summary><b>3. Tell other people that the extension is on you</b></summary>

Move your card into the **Assigned** column and comment on the issue so that everyone knows that it's on you.
</details>

<details>
<summary><b>4. Be sure tests pass</b></summary>

First of all, check that tests pass up to v2.11 by checking the CI status or launching this command from your machine:

```bash
$ rm -rf Gemfile.lock spec/dummy
$ SOLIDUS_BRANCH=v2.9 bundle
$ SOLIDUS_BRANCH=v2.9 bundle exec rake

$ rm -rf Gemfile.lock spec/dummy
$ SOLIDUS_BRANCH=v2.10 bundle
$ SOLIDUS_BRANCH=v2.10 bundle exec rake

$ rm -rf Gemfile.lock spec/dummy
$ SOLIDUS_BRANCH=v2.11 bundle
$ SOLIDUS_BRANCH=v2.11 bundle exec rake
```

Once all the above tests pass, move it into the **Specs Pass up to 2.11** column.
</details>

<details>
<summary><b>5. Remove all deprecation warnings</b></summary>

While running specs against Solidus v2.11 with

```bash
$ rm -rf Gemfile.lock spec/dummy
$ SOLIDUS_BRANCH=v2.11 bundle
$ SOLIDUS_BRANCH=v2.11 bundle exec rake
```

you should see some deprecation warnings. Change the extension code so that those warnings are no more present, being sure the specs always pass against the other Solidus versions as well.

Once all deprecation warnings are removed, move the card into **Have no deprecated code** column.

At this point tests should also pass against Solidus 3.0 (master). You can verify that with:

```bash
$ rm -rf Gemfile.lock spec/dummy
$ bundle
$ bundle exec rake
```

Push your code to a PR 

</details>

<details>
<summary><b>6. Create a PR</b></summary>

Push your code into a branch of your fork and create a PR against the main extension repository. At this point, you should see the CircleCI jobs passing and each step will be similar to:

![app circleci com_pipelines_github_solidusio_solidus_stripe_360_workflows_8b74af97-c13d-4c11-bb10-4ca4d70c6562_jobs_852 (1)](https://user-images.githubusercontent.com/167946/106292412-12106400-624d-11eb-944b-cb074e9aaf5e.png)

This is an [example of the passing CI](https://app.circleci.com/pipelines/github/solidusio/solidus_stripe/360/workflows/8b74af97-c13d-4c11-bb10-4ca4d70c6562/jobs/852) for an extension that is already compatible with Soldius 3.0: [solidus_stripe](https://github.com/solidusio/solidus_stripe).

</details>

And that's it. Someone will review and merge your PR soon. 🎉 

#### Can I open multiple PRs to prepare an extension (eg. one for specs, one for deprecations)?

Yes, but please consider that your PRs will likely not merged immediately although we'll do our best to review PRs as soon as possible. If you make multiple PRs and one depends on the other(s), they could need some rebase work later once they'll be merged.

#### What if I'm not done by the end of the day with my extension?

It depends. If you have time to finish the work during the next few days (no more than 2) you can keep the issue assigned to yourself. Please, just comment on the issue with your intention to complete the work. If you think you won't finish the work for any reason, no worries, that's perfectly fine. We just ask you to let us know by commenting on the issue so that someone else will be able to pick it up later. 

#### Address name migration

In Solidus 3 (currently tracked in the GitHub master branch) we have a new way of representing the user's name related to the addresses. We gradually transitioned from having two different fields (`firstname` and `lastname`) to a single field (`name`). The rationale behind this choice can be found here: https://github.com/solidusio/solidus/issues/3234.

We need extensions to be able to work with both versions and to help transitioning their code we created a helper method into the `solidus_support` gem that understands which version is currently being used.  `solidus_support` should already be a runtime dependency of all the extensions. After being sure `solidus_support` is at least at `0.8.0` version, you can use `SolidusSupport.combined_first_and_last_name_in_address?` to branch the different code paths.

Here's an example taken from the `solidus_stripe` PR that makes that extension compliant with Solidus 3.0:

```diff
    def address_full_name
-      current_order.bill_address&.full_name || form_data[:recipient]
+      bill_address_name = SolidusSupport.combined_first_and_last_name_in_address? ? current_order.bill_address&.name : current_order.bill_address&.full_name
+      bill_address_name || form_data[:recipient]
    end
```

##### NoMethodError: undefined method `gsub' for nil:NilClass

```
rake aborted!
NoMethodError: undefined method `gsub' for nil:NilClass
/home/circleci/project/vendor/bundle/v2.10/ruby/2.5.0/gems/octokit-4.20.0/lib/octokit/repository.rb:14:in `from_url'
/home/circleci/project/vendor/bundle/v2.10/ruby/2.5.0/gems/solidus_dev_support-2.3.0/lib/solidus_dev_support/rake_tasks.rb:81:in `block in install_changelog_task'
/home/circleci/project/vendor/bundle/v2.10/ruby/2.5.0/gems/github_changelog_generator-1.15.2/lib/github_changelog_generator/task.rb:42:in `define'
/home/circleci/project/vendor/bundle/v2.10/ruby/2.5.0/gems/github_changelog_generator-1.15.2/lib/github_changelog_generator/task.rb:36:in `initialize'
/home/circleci/project/vendor/bundle/v2.10/ruby/2.5.0/gems/solidus_dev_support-2.3.0/lib/solidus_dev_support/rake_tasks.rb:79:in `new'
/home/circleci/project/vendor/bundle/v2.10/ruby/2.5.0/gems/solidus_dev_support-2.3.0/lib/solidus_dev_support/rake_tasks.rb:79:in `install_changelog_task'
/home/circleci/project/vendor/bundle/v2.10/ruby/2.5.0/gems/solidus_dev_support-2.3.0/lib/solidus_dev_support/rake_tasks.rb:26:in `install'
/home/circleci/project/vendor/bundle/v2.10/ruby/2.5.0/gems/solidus_dev_support-2.3.0/lib/solidus_dev_support/rake_tasks.rb:11:in `tap'
/home/circleci/project/vendor/bundle/v2.10/ruby/2.5.0/gems/solidus_dev_support-2.3.0/lib/solidus_dev_support/rake_tasks.rb:11:in `install'
/home/circleci/project/Rakefile:4:in `<top (required)>'
/home/circleci/project/vendor/bundle/v2.10/ruby/2.5.0/gems/rake-13.0.3/exe/rake:27:in `<top (required)>'
/usr/local/bin/bundle:30:in `block in <main>'
/usr/local/bin/bundle:22:in `<main>'
(See full trace by running task with --trace)
```

To fix this error just add this line to the extension's gemspec :

```ruby
s.metadata["source_code_uri"] = s.homepage if s.homepage
```

This is needed for the automatic changelog generator and here's an [example commit](https://github.com/solidusio-contrib/solidus_product_assembly/commit/16ea4c49cb2c54c04f7a5dc325a5f1c6aea04e18) that fixes it.

**Comment by peterberkenbosch** [comment]
Awesome! Thanks @kennyadsl! Let's do this!

**Comment by afdev82** [comment]
It's a good idea.
I would like to help.

**Comment by seand7565** [comment]
Wow, this is awesome documentation for this event! Looking forward to getting some extensions up-to-date! 🔧

**Comment by forkata** [comment]
This is great 🎉 Sign me up!

**Comment by michaelmichael** [comment]
I'll join in.

**Reply by kennyadsl** [reply]
Thanks, let me know how it goes!

**Comment by jarednorman** [comment]
I'm excited for this!

**Reply by jarednorman** [reply]
I think three our four Super Good devs are going to help out a little with this, perhaps this afternoon. I'll swing by towards the end of my day and see if I can knock a couple ones that are left over out too.

**Reply by seand7565** [reply]
Awesome! Just let me know their GH handle or have them comment here and I'll add them to the hackathon board!

**Comment by ccarruitero** [comment]
I would like to participate

**Reply by kennyadsl** [reply]
I signed you up!

**Comment by coralcable** [comment]
I'd like to contribute! 🚀 Sign me up on the board! 👍

**Reply by kennyadsl** [reply]
Done!

**Comment by MFRWDesign** [comment]
Can't wait to get started! I'm in! 🔥

**Reply by kennyadsl** [reply]
Added to the board!

**Comment by zabrador** [comment]
I'd love to help!

**Reply by seand7565** [reply]
🎉 Added to the board!

**Comment by AlistairNorman** [comment]
Hey, I'd like to join in. Can you add me to the board? Thanks!

**Reply by seand7565** [reply]
Done! 🎉

**Comment by benjaminwil** [comment]
Hey, could you add me to the board? I will try to do one today.

**Reply by seand7565** [reply]
Done! 😄

**Comment by nvandoorn** [comment]
Looking to get added to the board. Thanks!

**Reply by seand7565** [reply]
🔑 You have access!

**Comment by kennyadsl** [comment]
All the extensions are now compliant with Solidus 3.0. Thanks to anyone who contributed, especially to @cpfergus1 that resolved all the remaining ones during this time!
